### DISCUSSION TITLE
Are TPM and NOPM metrics affected by external transactions on SQL Server?

### DISCUSSION THREAD

**Discussion by qwerty1954** [TPC-Council/HammerDB#479] [Q&A] [answered]
Hello!

Question: Are the TPM and NOPM metrics on SQL Server influenced by transactions that take place outside of the test?

Background: I was running a benchmark between 2 identical SQL Server instances with the exception of 1 of them having an additional process on it that ran selects on the engine (not on the benchmark database, however). This process uses a login created specifically for it.

The results surprised me, as I'd expect the SQL Server instance with the additional process to have lower TPM/NOPM metrics, but the opposite occurred. This led me to wondering if the selects ran by the additional process impacted the metric calculation.

Looking at https://github.com/TPC-Council/HammerDB/discussions/197, it seems NOPM should not be impacted by the additional selects, as the metric looks to be queried from the test database itself. On the other hand, it seems like TPM for SQL Server is calculated based on some generic performance counter, so it might be impacted by the additional process.

If that is the case, are there any recommendations to get around this? I know that with timeseries enabled, we can get a breakdown of all the procedures carried out by the vusers. Maybe a manual calculation could be done?

Any insight or confirmation here would be greatly appreciated. Thanks, and happy holidays!

**Comment by sm-shaw** [comment, marked as answer]
Hi,

Being open source you can look at the queries to see exactly what is being run, so for SQL Server. 

NOPM = `select sum(d_next_o_id) from district`

TPM = `select cntr_value from sys.dm_os_performance_counters where counter_name = 'Batch Requests/sec'`

So it should be clear that NOPM is not affected by other queries, however as we are running a query on the schema being tested we only want to run this at the start and end of the tests and then calculate the new orders per minute by the difference between the two and the time it ran for.  Only a new order transaction will increase NOPM. 

TPM for SQL Server you will notice is exactly the same figure as you can see in Activity Monitor in SSMS, querying the performance counters does not impact the test so we can query this as much as we like, which is why the GUI shows TPM and not NOPM however of course any other workload on the same database will also result in a higher TPM.

A general description in presentations is to think of NOPM and TPM like the dials on a car NOPM is how fast you are going, TPM is how hard the engine is working - if you are running other queries the engine is working harder to cope with the extra work but you are not going any faster. 

Note the TPM/NOPM ratio is not fixed across all databases.

Also as open source you can modify/add transaction counts to include any metrics you like as part of the test by finding the statements above and adding similar.
